services/docker/docker-version.service.js

```diff
@@ -0,0 +1,127 @@
+const { NotFound } = require('..')
+
+const buildSchema = Joi.object({
+  count: nonNegativeInteger.required(),
+  results: Joi.array()
+    .items(
+      Joi.object({
+        name: Joi.string().required(),
+        images: Joi.array().items(
+          Joi.object({
+            digest: Joi.string(),
+            architecture: Joi.string().required(),
+          })
+        ),
+      }).required()
+    )
+    .required(),
+}).required()
+
+module.exports = class DockerVersion extends BaseJsonService {
+  static get category() {
+    return 'version'
+  }
+
+  static get route() {
+    return buildDockerUrl('version', true, true)
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Docker Image Version (latest by date)',
+        pattern: ':user/:repo',
+        namedParams: { user: '_', repo: 'alpine' },
+        staticPreview: this.render({ version: '3.9.5', sort: 'date' }),
+      },
+      {
+        title: 'Docker Image Version (latest SemVer)',
+        pattern: ':user/:repo',
+        namedParams: { user: '_', repo: 'alpine' },
+        queryParams: { sort: 'semver' },
+        staticPreview: this.render({ version: '3.11.3' }),
+      },
+      {
+        title: 'Docker Image Version (tag latest SemVer)',
+        pattern: ':user/:repo/:tag',
+        namedParams: { user: '_', repo: 'alpine', tag: '3.6' },
+        staticPreview: this.render({ version: '3.6.5' }),
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return {
+      label: 'version',
+      color: 'blue',
+    }
+  }
+
+  static render({ version }) {
+    return { message: version }
+  }
+
+  async fetch({ user, repo, page }) {
+    page = page ? `&page=${page}` : ''
+    return this._requestJson({
+      schema: buildSchema,
+      url: `https://registry.hub.docker.com/v2/repositories/${getDockerHubUser(
+        user
+      )}/${repo}/tags?page_size=100&ordering=last_updated${page}`,
+      errorMessages: { 404: 'image or tag not found' },
+    })
+  }
+
+  async handle({ user, repo, tag }, { sort }) {
+    let data = []
+    let version = ''
+
+    if (!tag && sort === 'date') {
+      data = await this.fetch({ user, repo })
+    } else {
+      data = await getMultiPageData({
+        user,
+        repo,
+        fetch: this.fetch.bind(this),
+      })
+    }
+
+    if (!tag && sort === 'date') {
+      version = data.results[0].name
+      if (version === 'latest') {
+        const { digest } = data.results[0].images.find(
+          i => i.architecture === 'amd64'
+        ) // Digest is the unique field that we utilise to match images
+        data = await getMultiPageData({
+          user,
+          repo,
+          fetch: this.fetch.bind(this),
+        })
+        version = getDigestSemVerMatches({ data, digest })
+      }
+    } else if (!tag && sort === 'semver') {
+      const matches = data.map(d => d.name)
+      version = latest(matches)
+    } else {
+      version = data.find(d => d.name === tag)
```

Comment:
i'm a little confused by what's happening in this `else` block so a few questions to help me get a better understanding. 

I believe there's 3 `tag` parameter value scenarios that we'd encounter here:
1. User specifies `latest` for the badge `tag` route param
2. User specifies a value for the badge `tag` route param that's a match (partial or exact) for multiple tags, for example `../_/alpine/3.9`
3. User specifies a value for the badge `tag` route param that's a match for exactly one tag. 

In scenario 1, are we just using the digest from the `latest` tag to search for the corresponding tag in the full tag list based on the digest? 

In scenario 2, do we just need to find all the tag names that are fuzzy matches to the user-provided `tag` route param? For example with the `../_/alpine/3.9` badge route, there's two 3.9* like tags (`3.9` and `3.9.5`) so we should be selecting `3.9.5` as the semver-latest value for that subset of tags, correct? Is a digest lookup necessary, and if so why?

Reply:
The digest lookup would be necessary for scenarios where someone passes in a tag that doesn't conform to traditional SemVer. For example if you try `../_/node/alpine`.

Here's the data array based on the `digest` matching for said lookup:
```
[
  'alpine3.11',
  'alpine',
  '13.7.0-alpine3.11',
  '13.7.0-alpine',
  '13.7-alpine3.11',
  '13.7-alpine',
  '13-alpine3.11',
  '13-alpine'
]
```

It also looks if we use the helper to determine the latest SemVer on this, it returns `alpine3.11`.
I do not believe that's necessarily providing the right information in this scenario, I think the correct tag to display is `13.7.0-alpine3.11`.

This worked with the original very simple matching logic from an earlier commit. I've just a PR for this now.

Reply:
> I think the correct tag to display is 13.7.0-alpine3.11

How are we making that determination? To be honest, just looking at that set of tags I've no idea what qualifies as the most recent (by semver) tag in that set.

Reply:
Also it's worth calling out that if someone passes in an explicit tag other than `latest`, for example: `nightly`, `development`, `master` or whatever other kind of variation that's may not be a SemVer notation itself, the `digest` lookup will identify if said tag has an associated `SemVer` notation, the `ubuntu` library repo and many others utilise these types of tags.

Reply:
> > I think the correct tag to display is 13.7.0-alpine3.11
> 
> How are we making that determination? To be honest, just looking at that set of tags I've no idea what qualifies as the most recent (by semver) tag in that set.

I'd say without any context the closest thing to a full semver notation (Major, Minor, Patch) is the third and fourth tag and I can just eyeball that. The only difference between the two being the third shows the actual version of alpine that is being utilised.

Looking at all the various different library repos, it seems when they are notating with semver that they are doing so properly and seperating with periods `.`.

Another example data set for `latest` on `ubuntu` is:
```
[ 
  'latest',
  'bionic-20200112',
  'bionic',
  '18.04'
]
```
With `18.04` being selected here.

Or we have `golang` with both `latest` and `alpine`
```
[
  'latest',
  'buster',
  '1.13.7-buster',
  '1.13.7',
  '1.13-buster',
  '1.13',
  '1-buster',
  '1'
]
```
with `1.13.7` being selected here.
```
[
  'alpine3.11',
  'alpine',
  '1.13.7-alpine3.11',
  '1.13.7-alpine',
  '1.13-alpine3.11',
  '1.13-alpine',
  '1-alpine3.11',
  '1-alpine'
]
```
with `1.13.7-alpine3.11` being selected here.

Reply:
Ah okay. That makes a bit more sense now 👍